src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,258 @@
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        if (generationsLeft > 0)
+            generationsLeft -= 1;
+
+        if (generationsLeft > 0)
+            MarkPatches(totalTimePassed);
+
+        if (hasEventAlreadyHappened)
+            return;
+
+        if (generationsLeft == -1)
+        {
+            TryToTriggerEvent(totalTimePassed);
+        }
+        else if (generationsLeft == 0)
+        {
+            FinishEvent();
+        }
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (patch.Depth[0] == 0)
+            {
+                patch.AddPatchEventRecord(WorldEffectVisuals.GlobalGlaciation, totalTimePassed);
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        eventDuration = random.Next(Constants.GLOBAL_GLACIATION_MIN_DURATION, Constants.GLOBAL_GLACIATION_MAX_DURATION);
+        generationsLeft = eventDuration;
+
+        foreach (var (index, patch) in targetWorld.Map.Patches)
+        {
+            if (patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave)
+            {
+                ChangePatchProperties(index, patch, totalTimePassed);
+            }
+        }
+
+        LogBeginningOfGlaciation();
+    }
+
+    private bool AreConditionsMet()
+    {
+        var numberOfSurfacePatches = 0;
+        var patchesExceedingOxygenLevel = 0;
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (patch.Depth[0] != 0 || patch.BiomeType == BiomeType.Cave)
+                continue;
+
+            patch.Biome.TryGetCompound(Compound.Oxygen, CompoundAmountType.Biome, out var oxygenLevel);
+            if (oxygenLevel.Ambient >= Constants.GLOBAL_GLACIATION_OXYGEN_THRESHOLD)
+                patchesExceedingOxygenLevel += 1;
+
+            numberOfSurfacePatches += 1;
+        }
+
+        // Just prevent dividing by zero, but that shouldn't be possible anyway
+        numberOfSurfacePatches = numberOfSurfacePatches == 0 ? 1 : numberOfSurfacePatches;
+
+        return (float)patchesExceedingOxygenLevel / numberOfSurfacePatches >=
+            Constants.GLOBAL_GLACIATION_PATCHES_THRESHOLD
+            && random.NextFloat() <= Constants.GLOBAL_GLACIATION_CHANCE;
+    }
+
+    private void ChangePatchProperties(int index, Patch patch, double totalTimePassed)
+    {
+        modifiedPatchesIds.Add(index);
+        AdjustBackground(patch);
+        AdjustEnvironment(patch);
+        AddIceChunks(patch);
+        LogEvent(patch, totalTimePassed);
+    }
+
+    private void AdjustBackground(Patch patch)
+    {
+        patch.BiomeTemplate.Background = Background;
+        patch.CurrentSnapshot.Background = Background;
```

Comment:
Modifying the current snapshot is the safe operation, but modifying the template should be absolute not done here.